Lots of changes. Values and objects, not expressions, not "evaluations". (#167)
brevzin/cpp_proposals: 2996_reflection/reflection.md

Additionally, Bloomberg has open sourced a fork of Clang which provides a second implementation of this proposal, also available on Compiler Explorer (again thank you, Matt Godbolt), which can be found here: [https://github.com/bloomberg/clang-p2996](https://github.com/bloomberg/clang-p2996).

> Neither implementation is complete, but all significant features proposed by this paper have been implemented by at least one implementation (including namespace and template splicers). Both implementations have thier "quirks" and continue to evolve alongside this paper.

Reviewer: Typo
```suggestion
Neither implementation is complete, but all significant features proposed by this paper have been implemented by at least one implementation (including namespace and template splicers). Both implementations have their "quirks" and continue to evolve alongside this paper.
```

> |       `^` `::`
> |       `^` _namespace-name_
> |       `^` _type-id_
> |       `^` _id-expression_

The expression `^::` evaluates to a reflection of the global namespace. When the operand is a _namespace-name_ or _type-id_, the resulting value is a reflection of the designated namespace or type.

When the operand is an _id-expression_, the resulting value is a reflection of the designated entity found by lookup. This might be any of:

- a variable, static data member, or structured binding
- a function or member function

> - a nonstatic data member

Reviewer: ```suggestion
- a non-static data member
```

- a template or member template

> - an enumerator

Reviewer: Is the goal to be exhaustive here? So this should include type right?

For all other operands, the expression is ill-formed. In a SFINAE context, a failure to substitute the operand of a reflection operator construct causes that construct to not evaluate to constant.

> Earlier revisions of this paper allowed for taking the reflection of any _cast-expression_ that could be evaluated as a constant expression, as we believed that a constant expression could be internally "represented" by just capturing the value to which it evaluated. However, the possibility of side effects from constant evaluation (introduced by this very paper) renders this approach infeasible: even a constant expression would have to be evaluated every time it's spliced. It was ultimately decided to defer all support for expression reflection to a future paper.

Reviewer: I think we should note that we intend the syntax for that to be `^(expression)` so that it's clear we're looking ahead?

Author: Done.

The caret already has a meaning as a binary operator in C++ ("exclusive OR"), but that is clearly not conflicting with a prefix operator.
In C++/CLI (a Microsoft C++ dialect) the caret is also used as a new kind of `$ptr-operator$` ([dcl.decl.general]{.sref}) to declare ["handles"](https://learn.microsoft.com/en-us/cpp/extensions/handle-to-object-operator-hat-cpp-component-extensions?view=msvc-170).
That is also not conflicting with the use of the caret as a unary operator because C++/CLI uses the usual prefix `*` operator to dereference handles.

Apple also uses the caret in [syntax "blocks"](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/WorkingwithBlocks/WorkingwithBlocks.html) and unfortunately we believe that does conflict with our proposed use of the caret.

> Since the syntax discussions in SG7 landed on the use of the caret, new basic source characters have become available: `@`, `` ` ``{.op}, and `$`{.op}. While we have since discussed some alternatives (e.g., `@` for lifting, `\` and `/` for "raising" and "lowering"), we have grown quite fond of the existing syntax and don't feel that these alternatives offer sufficient value to justify re-painting the bikeshed.

Author: I buy that - removed everything following "we have grown fond of the existing syntax."

SG-7 eventually agreed to adopt the `[: ... :]` syntax --- with disambiguating tokens such as `typename` where needed --- which is a little lighter and more distinctive.

We propose `[:` and `:]` be single tokens rather than combinations of `[`, `]`, and `:`.
Among others, it simplifies the handling of expressions like `arr[[:refl():]]`.
On the flip side, it requires a special rule like the one that was made to handle `<::` to leave the meaning of `arr[::N]` unchanged and another one to avoid breaking a (somewhat useless) attribute specifier of the form `[[using ns:]]`.

A syntax that is delimited on the left and right is useful here because spliced expressions may involve lower-precedence operators.
However, there are other possibilities.
For example, now that `$`{.op} is available in the basic source character set, we might consider `@[$]{.op}@<$expr$>`.
This is somewhat natural to those of us that have used systems where `$`{.op} is used to expand placeholders in document templates.  For example:

::: std
```c++
@[$]{.op}@select_type(3) *ptr = nullptr;
```

> :::

Author: Fair enough - I'm weakly opposed to re-opening the syntax discussion, but I've reverted the change.

I don't personally feel the need to space around `[:R:]`, but I see your point.

The splicers described above all take a single object of type `std::meta::info` (described in more detail below).
However, there are many cases where we don't have a single reflection, we have a range of reflections - and we want to splice them all in one go.
For that, the predecessor to this paper, [@P1240R0], proposed an additional form of splicer: a range splicer.

Construct the [struct-to-tuple](#converting-a-struct-to-a-tuple) example from above. It was demonstrated using a single splice, but it would be simpler if we had a range splice:

::: cmptable

> #### With Single Splice

Reviewer: No 3 `#`s is correct, that's what `cmptable` is for as a heading.

Author: Thanks, done.

Among others, it simplifies the handling of expressions like `arr[[:refl():]]`.
On the flip side, it requires a special rule like the one that was made to handle `<::` to leave the meaning of `arr[::N]` unchanged and another one to avoid breaking a (somewhat useless) attribute specifier of the form `[[using ns:]]`.

> A syntax that is delimited on the left and right is useful here because spliced expressions may involve lower-precedence operators.

Reviewer: Let's also add this sentence which I thought we had somewhere but we don't. 
```suggestion
A syntax that is delimited on the left and right is useful here because spliced expressions may involve lower-precedence operators. Additionally, it's important that the left- and right-hand delimiters are different so as to allow nested splices when that comes up. 
```

  - any function or member function
  - any variable, static data member, or structured binding
  - any non-static data member
  - any enumerator
  - any template
  - any namespace (including the global namespace) or namespace alias
  - any object that is a _permitted result of a constant expression_

>   - any value with _structural type_ (with the caveat that any references or pointers must designate objects that are permitted results of constant expressions)

Reviewer: Use the same term:
```suggestion
- any value with _structural type_ that is a permitted result of a constant expression
```

Actually why do we need the requirement on structural? If we end up passing this value into `substitute` then we'll pick up that requirement on substitution, but like... what's wrong with:
```cpp
class C {
    int i; // private
public:
    constexpr int get() const { return i; }
};

constexpr auto r = meta::reflect_result(C{});
static_assert([: r :].get() == 0);
```

Author: Whatever class of things we allow here, the compiler will have to know how to compare. I think we were trying to rule out custom comparison operators, but I don't think "structural" does that for us anyway.

@daveedvdv Any thoughts on what the right class is here? Aggregate < Structural < Literal, AFAICT

Author: (applied the suggested change)

The second overload behaves the same as the first overload, except instead of evaluating `F(A@~0~@, A@~1~@, ..., A@~N~@)`, we require that `F` be a reflection of a template and evaluate `F<T@~0~@, T@~1~@, ..., T@~M~@>(A@~0~@, A@~1~@, ..., A@~N~@)`. This allows evaluating `reflect_invoke(^std::get, {reflect_value(0)}, {e})` to evaluate to, approximately, `^std::get<0>([: e :])`.

### `reflect_result<T>`

:::bq
```c++
namespace std::meta {
  template<typename T> consteval auto reflect_result(T expr) -> info;
}
```
:::

If `expr` does not have structural type, then `reflect_result(expr)` fails to be a constant expression.

> Otherwise, if `T` is of reference or pointer type, or for each subobject of `expr` having reference or pointer type if `T` is of class type, if the reference or pointer value designates an entity that is not a _permitted result of a constant expression_ ([expr.const]), then `reflect_result(expr)` fails to be a constant expression.

Reviewer: I think you just want to say that `expr` is a permitted result.

Otherwise, `reflect_result(expr)` produces a reflection of the result of `static_cast<T>(expr)`.

### `extract<T>`

:::bq
```c++
namespace std::meta {
  template<typename T> consteval auto extract(info) -> T;
}
```
:::

If `r` is a reflection for a value of type `T`, `extract<T>(r)` is a prvalue whose evaluation computes the reflected value.

If `r` is a reflection for an object of non-reference type `T`, `extract<T&>(r)` and `extract<T const&>(r)` are lvalues referring to that object.
If the object is usable in constant expressions [expr.const], `extract<T>(r)` evaluates to its value.

If `r` is a reflection for an object of reference type `T` usable in constant-expressions, `extract<T>(r)` evaluates to that reference.

> If `r` is a reflection for a function of type `R(A_0, ... A_n)`, `extract<R(*)(A_0, ..., A_n)>(r)` evaluates to a pointer to that function.

Reviewer: Just for sanity (and also want to cover `noexcept`, etc.)
```suggestion
If `r` is a reflection for a function of type `F`, `extract<F*>(r)` evaluates to a pointer to that function.
```

::: addu
**The Reflection Operator   [expr.reflect]**

[#]{.pnum} The unary `^` operator (called _the reflection operator_) produces a prvalue --- called a _reflection_ --- whose type is the reflection type (i.e., `std::meta::info`).
That reflection represents its operand.

[#]{.pnum} Every value of type `std::meta::info` is either a reflection of some operand or a *null reflection value*.

[#]{.pnum}

::: example
```

> static_assert(is_type(^int()));    // ^ applies to the type-id "int()"

Reviewer: This example is kind of hanging now, since it used to describe the ambiguity. But still useful to have... somewhere. I guess worry about it later.

:::

[#]{.pnum} When applied to `::`, the reflection operator produces a reflection for the global namespace.
When applied to a `$namespace-name$`, the reflection operator produces a reflection for the indicated namespace or namespace alias.

[#]{.pnum} When applied to a `$template-name$`, the reflection operator produces a reflection for the indicated template.

[#]{.pnum} When applied to a `$concept-name$`, the reflection operator produces a reflection for the indicated concept.

[#]{.pnum} When applied to a `$type-id$`, the reflection operator produces a reflection for the indicated type or type alias.

> [#]{.pnum} When applied to an `$id-expression$` ([expr.prim.id]{.sref}), the reflection operator produces a reflection of the variable, function, enumerator constant, or nonstatic member designated by the operand.

Reviewer: We hyphenate non-static
```suggestion
[#]{.pnum} When applied to an `$id-expression$` ([expr.prim.id]{.sref}), the reflection operator produces a reflection of the variable, function, enumerator constant, or non-static member designated by the operand.
```

consteval bool is_class_template(info r);
consteval bool is_alias_template(info r);
consteval bool is_concept(info r);
consteval bool is_structured_binding(info r);
consteval bool is_value(info r);
```
[#]{.pnum} *Returns*: `true` if `r` designates a function template, class template, variable template, alias template, concept, structured binding, or value respectively. Otherwise, `false`.

```cpp
consteval bool is_object(info r);
```

> [#]{.pnum} *Returns*: `true` if `r` designates an object. Otherwise, `false`.

Reviewer: Why split this one out?

[#]{.pnum} *Mandates*: `r` is a reflection designating either a class type or a namespace and `(std::predicate<Fs, info> && ...)` is `true`.

> [#]{.pnum} *Returns*: A `vector` containing the reflections of all the direct members `m` of the entity, not including any structured bindings, designated by `r` such that `(filters(m) && ...)` is `true`.

Reviewer: ```suggestion
[#]{.pnum} *Returns*: A `vector` containing the reflections of all the direct members `m` of the entity, excluding any structured bindings, designated by `r` such that `(filters(m) && ...)` is `true`.
```
